presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -807,6 +809,18 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
             ImmutableList<TypeSignatureProvider> argumentTypes = argumentTypesBuilder.build();
             Signature function = resolveFunction(node, argumentTypes, functionRegistry);
 
+            if (node.getOrderBy().isPresent()) {
+                if (function.getKind() == SCALAR) {
+                    throw new SemanticException(MUST_BE_AGGREGATION_FUNCTION, node, "ORDER BY is only valid for aggregation functions");
```

Comment:
Hmm, ok. AggregationAnalyzer.Visitor#visitFunctionCall is only called for aggregation functions (from StatementAnalyzer.analyzeAggregations). The trick was to consider all functions with filter / window as aggregation function. I can do it that way too. It wasn't intuitive to me.